python-package/lightgbm/plotting.py

```diff
@@ -34,7 +34,7 @@ def _float2str(value: float, precision: Optional[int]) -> str:
 
 def plot_importance(
     booster: Union[Booster, LGBMModel],
-    ax=None,
+    ax: Optional[Any] = None,
```

Comment:
Shouldn't these be `plt.Axes`?

Reply:
Ok yeah, I did actually try a couple versions of this. Let me share what I was thinking, see if you can think of something better.

So my first thought was this:

```python
if TYPE_CHECKING:
    import matplotlib

...

def plot_importance(
    booster: Union[Booster, LGBMModel],
    ax: "Optional[matplotlib.axes.Axes]" = None,
    ...
```

That works, but I think it'd then make `matplotlib` a required dependency for type-checking `lightgbm` (and by extension, anything depending on `matplotlib`, like `mlforecast` or `optuna`). I thought that was a heavy dependency to force on other projects just for type-checking of plotting code here.

So then I tried this:

```python
if TYPE_CHECKING:
    from typing import TypeAlias
    if MATPLOTLIB_INSTALLED:
        import matplotlib
        matplotlib_axes: TypeAlias = matplotlib.axes.Axes
    else:
        matplotlib_axes: TypeAlias = Any

...

def plot_importance(
    booster: Union[Booster, LGBMModel],
    ax: "Optional[matplotlib_axes]" = None,
    ...
```

But `mypy` did not like that:

```text
python-package/lightgbm/plotting.py:28: error: Cannot assign multiple types to name "matplotlib_axes" without an explicit "Type[...]" annotation  [misc]
python-package/lightgbm/plotting.py:28: error: Name "matplotlib_axes" already defined on line 26  [no-redef]
```

And then at that point I gave up and just went with `Optional[Any]`, to at least enable these stricter settings in `mypy`.

Can you think of a better way to do this? Is it worth moving the rest of the `matplotlib` imports into `compat.py` like we do with `dask`?

e.g.

https://github.com/microsoft/LightGBM/blob/62a5a161e09df174c19b5de75db6878e0cf88b89/python-package/lightgbm/compat.py#L150-L152

https://github.com/microsoft/LightGBM/blob/62a5a161e09df174c19b5de75db6878e0cf88b89/python-package/lightgbm/dask.py#L59-L63

Reply:
> That works, but I think it'd then make matplotlib a required dependency for type-checking lightgbm

I'm not sure about that, since we have this try/except here

https://github.com/microsoft/LightGBM/blob/dd31208ab7a7aea86762830697b00666f843ded9/python-package/lightgbm/__init__.py#L16-L19

so if the code being checked isn't directly using one of those functions I don't think matplotlib will be required. And if it is then it probably has it installed and be thorough with the type checks.

I just ran mypy on this file (from this branch but adding your first suggestion):
```python
import lightgbm as lgb

def f(bst: lgb.Booster) -> None:
    lgb.plot_importance(bst, ax=1)
```
with and without matplotlib installed. When it isn't installed it didn't give me any errors and when it was installed it did give me the expected
`error: Argument "ax" to "plot_importance" has incompatible type "int"; expected "Axes | None"  [arg-type]`

Reply:
oh wow, nice!! Ok thank you for testing that. I forgot about that try-catched import in `__init__.py`.

Pushed that more informative hint in https://github.com/microsoft/LightGBM/pull/6339/commits/2bdede9fa9547c5bc43e3d6001602f6fad9f6696